### ISSUE TITLE
Serial protocol

### ISSUE THREAD

**Issue by gcielniak** [LCAS/lcas_hardware#2]
@larsgrim We are developing some hardware for Thorvald such as penetrometer and I was thinking about a serial communication standard and protocol that could potentially be shared by all external low-bandwidth devices.  Have you had any thoughts on that?
As I see alternatives are as follows:
 - RS232: simple hardware and interconnectivity, a bit dated, perhaps not too robust and I am not aware of any good higher-level protocols for I/O devices.
- CAN (CANOpen): open standard, most likely requiring beefier microcontrollers;
- Ethernet/IP: Ethernet-based, automation standard based on CIP protocol. As I understand it typically requires 32bit controllers and is a bit complex, although we managed to interface one of the motor controllers for the old penetrometer.

Physical connections are also important to consider. It might be that there is no universal choice but thought to run that through you and see what you think.

**Comment by larsgrim** [comment]
It would be good with a standard for low-bandwidth devices, but it may be better with a standard that goes with everything.

I suppose future implements will require far more bandwidth and probably also more processing power than our current implements. Then I guess a separate computer on the implement and Ethernet communication between Implement/Thorvald would make more sense.

I recently added some support for ROS/CAN communication for the existing CAN on the robot. CAN is a bit dated, but it is good for low-bandwidth devices. It is simple to connect more than one device at once, and Thorvald's CAN bus is easy to connect to. It may be OK for the moment.

Message definition for CAN frames were added in https://github.com/LCAS/Thorvald/commit/d063444283336ba6b26ccc211071ea131b6953dd and topics for CAN frames in https://github.com/LCAS/Thorvald/commit/e59fa9b15092af3b77c5f5a0c5ebbd1dba68432a

Received CAN messages that are not from Thorvald's prop. or steer. motors are published to `/can_frames_device_r` by `base_driver` so that they are available for other nodes.

To send messages to a CAN device, publish to `/can_frames_device_t` and `base_driver`will get them on the bus.

The implement can then be a controlled by a node publishing and subscribing to these topics. `base_driver` sends/receives messages to/from CAN.

There is a service for switching this communication off and on currently named `/mute_extra_can_device_commands`.


I have not used microcontrollers that support CAN, but there are some to chose from. There are also CAN shields for Arduinos and such for prototyping. <a href=http://www.roboteq.com>Roboteq</a> motor controllers support CAN (CANOpen protocol), they have inputs and outputs and can be programmed. I like these and have used them quite a lot. We will use a Roboteq as the "brain" for our sprayer. This controller will get input from pressure sensors and it will control valves etc.





What do you think?

**Comment by gcielniak** [comment]
Yes, very good and it looks like there is a lot of infrastructure supporting CAN already which should make interfacing simple. We will worry about an universal solution supporting also high-bandwidth devices a bit later. Cheers!
